devito/arguments.py

```diff
@@ -1,152 +1,133 @@
 import abc
 
 import numpy as np
-from cached_property import cached_property
-from collections import namedtuple
+from collections import OrderedDict
+from functools import reduce
+from itertools import chain
 
 from devito.exceptions import InvalidArgument
-from devito.logger import debug
-
-""" This module provides a set of classes that help in processing runtime arguments for
-    kernels generated by devito. There are two class hierarchies here:
-    - ArgumentProvider: These are for objects that might be used in the expression
-      provided to the operator as symbols but might resolve to runtime arguments after
-      code generation. Each ArgumentProvider provides one (or more) Argument
-      object(s).
-    - Argument: Classes inheriting from this are for objects that represent the
-      argument itself. Each ArgumentProvider might provide one or more such objects
-      which are used as placeholders for the argument as well as for verification and
-      derivation of default values.
+from devito.tools import filter_ordered, flatten, GenericVisitor
+from devito.function import CompositeFunction, SymbolicFunction
+from devito.dimension import Dimension
+from devito.ir.support.stencil import retrieve_offsets
+
+
+""" This module contains a set of classes and functions to deal with runtime arguments
+to Operators. It represents the arguments and their relationships as a DAG (N, E) where
+every node (N) is represented by an object of :class:`Parameter` and every edge is an
+object of class :class:`Dependency`.
+
+The various class hierarchies are explained here:
+Parameter: Any node of the dependency graph has to necessarily be of this type. The node
+           may or may not represent an actual runtime argument passed to the kernel.
+Argument: Subclass of Parameter. This represents a node in the dependency graph directly
+          corresponding to a runtime argument passed to the kernel.
+
+             Parameter
+                 |
+        -------------------
+DimensionParameter        |
+                          |
+                       Argument
+                          |
+                          |
+                --------------------
+                |         |        |
+        ScalarArgument    |        |
+                    TensorArgument |
+                                PtrArgument
+
+ArgumentEngine: The main external API for this module. It encapsulates all the argument
+                derivation and verification logic.
+
+ArgumentVisitor: Visits objects of devito's data types to return appropriate objects
```

Comment:
"Visits objects of devito's data types ... "

you mean ``:class:`function.Basic` `` ?

Reply:
Fixed